### PR TITLE
Sprite Updates

### PR DESCRIPTION
**Pull request by Pryme8**
- Fixed the spriteManager JSON loading now supports JSON string and Standard Objects with the frame structure both of an object and array type.
- Introduction of spriteMap to handle grid deployments of spriteAtlas
- Sample spriteMap scene added to playground
- adding resources to the playground to support additional spriteMap samples.

### CODE REVIEW

**Comment by deltakosh** [comment]
Are these resources free of right?

File: src/Sprites/spriteMap.ts [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+import { ActionManager } from "../Actions/actionManager";
+import { PickingInfo } from "../Collisions/pickingInfo";
+
+import "../Meshes/Builders/planeBuilder";
+
+/**
+ * Defines the minimum interface to fulfill in order to be a sprite map.
+ */
+export interface ISpriteMap extends IDisposable {
+
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: seems empty :)

File: src/Sprites/spriteMap.ts [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+    this._material.setTextureArray("tileMaps", this._tileMaps);
+    this._material.setTexture("animationMap", this._animationMap);
+    this._material.setFloat('time', this._time);
+
+    this._output = Mesh.CreatePlane(name + ":output", 1, scene, true);
+    this._output.scaling.x = this.options.outputSize.x;
+    this._output.scaling.y = this.options.outputSize.y;
+
+    let obfunction = null;
+       if (this._editable) {
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: weird tab issue

File: src/Sprites/spriteMap.ts [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+                this._time += this._scene.getEngine().getDeltaTime() * 0.01;
+                this._material.setFloat('time', this._time);
+                this._material.setVector2('mousePosition', this.getMousePosition());
+                this._material.setFloat('curTile', this.editorTile);
+            };
+
+            this._output.actionManager = new ActionManager(this._scene);
+            const drawing = () => {
+                let pos = this.getTileID();
+                console.log(pos);
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: need to remove all console.log

File: src/Sprites/spriteMap.ts [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+                if (pos.x == this.editorPos.x &&
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: ===

File: src/Sprites/spriteMap.ts [hunk 5]
```diff
@@ -0,0 +1,627 @@
...
+                            this.editorLayer = Math.max(0.0, this.editorLayer - 1) ;
+                        break;
+                        case ']':
+                            this.editorLayer = Math.min(this.editorLayer + 1, this.options.layerCount);
+                        break;
+                    }
+                })
+            );
+
+        }else {
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: space before else

File: src/Sprites/spriteMap.ts [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+            obfunction = () => {
+                this._time += this._scene.getEngine().getDeltaTime() * 0.01;
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: why 0.01?

File: src/Sprites/spriteMap.ts [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+        Texture.NEAREST_NEAREST,
+        Engine.TEXTURETYPE_FLOAT
+        );
+
+        return t;
+    }
+
+    /**
+    * Creates the tileMap texture Buffer
+    * @param buffer normaly and array of numbers, or a false to generate from scratch
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: typos

File: Playground/scripts/spriteMap.js [hunk 8]
```diff
@@ -0,0 +1,648 @@
...
+    // Load the SpriteSheet Associated with the JSON Atlas.
+    let spriteSheet = new BABYLON.Texture('./textures/spriteMap/none_trimmed/Legends_Level_A.png', scene,
+    false, //NoMipMaps
+    false, //InvertY usuaslly false if exported from TexturePacker
+    0, //Sampling Mode
+    null, //Onload, you could spin up the sprite map in a function nested here
+    null, //OnError
+    null, //CustomBuffer
+    false, //DeleteBuffer
+    5 //ImageFormageType RGBA
```
Comment (deltakosh) [inline Playground/scripts/spriteMap.js]: use constants and not numbers :) (same for sampling mode)

File: Playground/scripts/spriteMap.js [hunk 9]
```diff
@@ -0,0 +1,648 @@
...
+        stageSize: backgroundSize,
+        maxAnimationFrames:8,
+        baseTile : 42,
+        layerCount: 2,
+        flipU: true, //Sometimes you gotta flip the depending on the sprite format.
+        colorMultiply : new BABYLON.Vector3(0.3,0.3,0.3)
+    },
+    scene);    
+    
+    //Set all the available tiles to the top left corner of the background for Visual debugging, and refrence.
```
Comment (deltakosh) [inline Playground/scripts/spriteMap.js]: typos

File: src/Shaders/spriteMap.fragment.fx [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+					frameID = animationData.x;
+					break;
+				}
+				animationData = texture(animationMap, vec2((frameID+0.5)/spriteCount, aFrameSteps*f), 0.); 
+			}
+		}
+    
+		//Get Animation Frame
+
+		mat4 frameData = getFrameData(frameID+0.5);
```
Comment (deltakosh) [inline src/Shaders/spriteMap.fragment.fx]: this is true for all the files: need a space before and after operators

File: src/Shaders/spriteMap.fragment.fx [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+	
+	#ifdef EDITABLE
+		if(tileID == floor(mousePosition*stageSize)){
+			
+			vec2 eUV = fract(tUV);
+            if(flipU==1.0){
+                eUV.y = 1.0 - eUV.y;
+            }
+						
+			mat4 eframeData = getFrameData(curTile+0.5);
```
Comment (deltakosh) [inline src/Shaders/spriteMap.fragment.fx]: tabs are probably wrong

File: src/Sprites/spriteMap.ts (hunk 1 above)
Comment (deltakosh) [reply src/Sprites/spriteMap.ts]: You also need to include the ts file for your shaders or they will be tree shacked by webpack

File: src/Sprites/spriteMap.ts [hunk 12]
```diff
@@ -0,0 +1,627 @@
...
+    public name : string;
+
+    /** The JSON file with the frame and meta data */
+    public atlasJSON : {frames: Array<any>, meta: object};
+
+     /** The systems Sprite Sheet Texture */
+    public spriteSheet : Texture;
+
+    /** Arguments passed with the Constructor */
+    public options : any;
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: Please provide an interface for the options

File: src/Sprites/spriteMap.ts [hunk 13]
```diff
@@ -0,0 +1,627 @@
...
+
+    /** Public Sprite Storage array, parsed from atlasJSON */
+    public sprites : Array<any>;
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: I would rather prefer having an interface for each any. It will be better for type analysis

File: src/Sprites/spriteMap.ts [hunk 14]
```diff
@@ -0,0 +1,627 @@
...
+    public set animationMap(v: RawTexture) {
+        let buffer = v!._texture!._bufferView;
+        let am = this.createTileAnimationBuffer(buffer);
+        this._animationMap.dispose();
+        this._animationMap = am;
+        this._material.setTexture('animationMap', this._animationMap);
+    }
+
+    /** gets Editor Tile*/
+    public get editorTile() {
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: see my comment about editor

File: src/Sprites/spriteMap.ts [hunk 15]
```diff
@@ -0,0 +1,627 @@
...
+    }
+    /**
+    * Modfies the data of the animationMap
+    * @param cellID is the Index of the Sprite
+    * @param _frame is the target Animation frame
+    * @param toCell is the Target Index of the next frame of the animation
+    * @param time is a value between 0-1 that is the trigger for when the frame should change tiles
+    * @param speed is a global scalar of the time varable on the map.
+    */
+    addAnimationToTile(cellID: number= 0, _frame: number= 0, toCell: number= 0, time: number= 0, speed: number = 1): void {
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: All the utilities function should be flag as private and not exposed publicly (keep the API simple)

File: src/Shaders/spriteMap.fragment.fx [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+		texture(frameMap, vec2(fX, fdStep*1.), 0.),
+		texture(frameMap, vec2(fX, fdStep*2.), 0.),
+		vec4(0.)
+	);
+}
+
+void main(){
+    vec4 color = vec4(0.);    
+    vec2 tileUV = fract(tUV);
+    if(flipU==1.0){
```
Comment (deltakosh) [inline src/Shaders/spriteMap.fragment.fx]: We try to avoid using if as much as possible (really slow on low end devices). 
Use #ifdef as much as possible (see other shaders)

File: src/Sprites/spriteMap.ts [hunk 17]
```diff
@@ -0,0 +1,627 @@
...
+        this._tileMaps[_layer] = t;
+        this._material.setTextureArray("tileMap", this._tileMaps);
+    }
+
+    /**
+    * Creates the animationMap texture Buffer
+    * @param buffer normaly and array of numbers, or a false to generate from scratch
+    * @returns RawTexture of the animationMap
+    */
+    createTileAnimationBuffer(buffer: any): RawTexture {
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: All the utilities function should be flag as private and not exposed publicly (keep the API simple)

**Comment by Pryme8** [comment]
Roger that, I will review all of your input in the morning and make the correct changes.  Thank you for your extensive review!

The Sprite graphics are of my own creation (besides the player ones and the single BG graphic which Ill delete so there is no question.)

File: src/Sprites/spriteMap.ts (hunk 6 above)
Comment (Pryme8) [reply src/Sprites/spriteMap.ts]: It just made it smooth for the animations it seemed like.

 I just deleted it they have access to a speed variable when the animations are created.

File: src/Sprites/spriteMap.ts (hunk 2 above)
Comment (Pryme8) [reply src/Sprites/spriteMap.ts]: I think resolved.

**Comment by Pryme8** [comment]
Oh yeah, and my thoughts on the editor are right in line with you.  It could completely be deployed by the user and I could make a scene demonstrating how to do that.   I left the #ifdef in the shader though so that way we leave that option open.

File: src/Shaders/spriteMap.fragment.fx [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+uniform float maxAnimationFrames;
+
+uniform sampler2D frameMap;
+uniform sampler2D tileMaps[LAYERS];
+uniform sampler2D animationMap;
+
+uniform vec3 colorMul;
+
+
+#ifdef EDITABLE
```
Comment (deltakosh) [inline src/Shaders/spriteMap.fragment.fx]: I'm not for letting stuff not used

File: src/Shaders/spriteMap.fragment.fx [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+            vec4 nc = texture(spriteSheet, tileUV * frameSize+offset);
+            float alpha = min(color.a + nc.a, 1.0);
+            vec3 mixed = mix(color.xyz, nc.xyz, nc.a);
+            color = vec4(mixed, alpha);
+        }	
+    }
+
+    color.xyz*=colorMul;
+
+    #ifdef EDITABLE
```
Comment (deltakosh) [inline src/Shaders/spriteMap.fragment.fx]: I prefer removing it all

File: src/Sprites/spriteMap.ts [hunk 20]
```diff
@@ -0,0 +1,659 @@
...
+    /**
+	 * ISpriteJSONSpriteFrame basic object of the source data
+	 */
+    sourceSize : ISpriteJSONSpriteSourceSize;
+}
+
+/**
+ * Defines the basic options interface of a JSON atlas.
+ */
+export interface ISpriteJSONAtlas{
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: Do you mind moving them to individual files?

File: src/Sprites/spriteMap.ts (hunk 1 above)
Comment (deltakosh) [reply src/Sprites/spriteMap.ts]: I do not see the import for shaders?

File: src/Sprites/spriteMap.ts [hunk 21]
```diff
@@ -0,0 +1,659 @@
...
+    * Structure of frames
+    *  "filename": "Falling-Water-2.png",
+    * "frame": {"x":69,"y":103,"w":24,"h":32},
+    * "rotated": true,
+    * "trimmed": true,
+    * "spriteSourceSize": {"x":4,"y":0,"w":24,"h":32},
+    * "sourceSize": {"w":32,"h":32}
+    * @returns RawTexture of the frameMap
+    */
+    private createFrameBuffer(): RawTexture {
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: private functions start with _

File: src/Sprites/spriteMap.ts [hunk 22]
```diff
@@ -0,0 +1,659 @@
...
+        return t;
+    }
+
+    /**
+    * Modifies the data of the tileMaps
+    * @param _layer is the ID of the layer you want to edit on the SpriteMap
+    * @param pos is the iVector2 Coordinates of the Tile
+    * @param tile The SpriteIndex of the new Tile
+    */
+    public changeTiles(_layer: number = 0, pos: any , tile: number = 0): void {
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: no any

File: src/Sprites/spriteMap.ts [hunk 23]
```diff
@@ -0,0 +1,659 @@
...
+        this._tileMaps[_layer] = t;
+        this._material.setTextureArray("tileMap", this._tileMaps);
+    }
+
+    /**
+    * Creates the animationMap texture Buffer
+    * @param buffer normally and array of numbers, or a false to generate from scratch
+    * @returns RawTexture of the animationMap
+    */
+    private createTileAnimationBuffer(buffer: any): RawTexture {
```
Comment (deltakosh) [inline src/Sprites/spriteMap.ts]: should be buffer: ArrayBuffer | boolean

**Comment by Pryme8** [comment]
I've got a weird linting error on the build now too now that I gotta figure out.

File: src/Sprites/spriteMap.ts (hunk 20 above)
Comment (Pryme8) [reply src/Sprites/spriteMap.ts]: which directory has an example with the interfaces split off into separate files so I can reference?

Comment (Pryme8) [reply src/Sprites/spriteMap.ts]: Nevermind I figured it out.

File: src/Sprites/spriteMap.ts (hunk 23 above)
Comment (Pryme8) [reply src/Sprites/spriteMap.ts]: I had to use Nullable<ArrayBufferView>

**Comment by Pryme8** [comment]
I'm struggling with the linting...  the build go through fine on my local with incremental builds and then it's kicking me out on a full build but with no hints as to why.

Any idea what would cause that?

**Comment by deltakosh** [comment]
image


Just run num run build to get a automatic fix

**Comment by Pryme8** [comment]
I tried that and was not getting that same dialog... hmmm.
Let me try again on a different command line, maybe it's my wrapper that's giving me issues.

**Comment by Pryme8** [comment]
When I run build this is the error I get:
[2019-12-02T16_44_19_954Z-debug.log](https://github.com/BabylonJS/Babylon.js/files/3912333/2019-12-02T16_44_19_954Z-debug.log)

What is different with my setup then yours I am wondering?  If I could get it to actually tell me what is going wrong I could fix it.

**Comment by deltakosh** [comment]
just make sure you ran npm install recently in the gulp folder

**Comment by Pryme8** [comment]
Yeah that did not seem to fix it, I am trying a couple more things.  It seems like a bunch of my firefox tests are failing so I know something is up.

**Comment by Pryme8** [comment]
Im going to try to do a clean npm install with a reinstall of all the -g dependencies and see if I can snip this problem out.

**Comment by Pryme8** [comment]
Is there anything else I need to revise on this before merging?

**Comment by deltakosh** [comment]
The what's new and we are good ;)

**Comment by Pryme8** [comment]
I have no clue how to do that! Haha, its gotta be in the docs somewhere let me do some reading.

**Comment by deltakosh** [comment]
Just update this file:
https://github.com/BabylonJS/Babylon.js/blob/master/dist/preview%20release/what's%20new.md

**Comment by deltakosh** [comment]
Ok this is fine! Now doc and samples ;)

**Comment by Pryme8** [comment]
Roger that!
